submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,300 @@
+## Proposal
+
+This enhancement proposes the following changes to make Globalnet implementation scale better.
+
+Egress traffic:
+
+* By default, every cluster will be assigned a configurable number of globalIPs which will be
+  used as egress IPs for any cross-cluster communication. Support for more than a single globalIP
+  at the cluster level is done to avoid ephemeral port exhaustion issues.
+* Provision to allocate a set of globalIPs at the namespace/project level which takes precedence
+  over the globalIPs allocated to the cluster. All Pods in the namespace will use the same globalIPs
+  as EgressIPs.
+* Provision to allocate globalIPs to a set of Pods in a selected namespace or cluster-wide.
+  Such Pods will use globalIPs that take precedence over any globalIPs allocated at the default
+  namespace level.
+
+Ingress traffic:
+
+* Instead of assigning a globalIP every ClusterIP Service in the cluster, only exported Services
+  will be assigned a globalIP. Globalnet internally creates a GlobalIngressIP CR for every
+  exported service.
+* For Headless Services, necessary ingress rules will be programmed to support connectivity to
+  the backend Pods. In a Headless Service use-case, every backend Pod will be assigned a unique
+  globalIP as an Ingress IP. Clients can discover the associated globalIP using Lighthouse and
+  can connect to the globalIP assigned to the backend pods. For Pods backed by Headless Services,
+  if the user does not explicitly request a globalIP via GlobalEgressIP CRD, the same globalIP
+  assigned to the Pods will be used for both Ingress as-well-as Egress. However, if user creates
+  a GlobalEgressIP that matches the namespace, or the backend pods of Headless Service, the globalIP
+  assigned via GlobalEgressIP takes precedence over individual Pod globalIPs as EgressIP.
+
+Both EgressIP as well as IngressIPs will be allocated from the same globalCIDR Pool used in
+the Cluster. Globalnet will not support an EgressIP/IngressIP that is outside of globalCIDR
+Pool.
+
+![GlobalnetTopology](./images/globalnet-topology.png)
+
+If the external application in a remote Cluster is interested in consuming the exported
+Service, it will have to reach the globalIP assigned to the exportedService and the response
+from the backend Pods will have the source as the globalIP of the exportedService.
+
+On the other hand, when the backend Pod tries to talk to the external application in remote
+Cluster, the external application will see that traffic is coming with a globalIP that is
+assigned either at the Cluster level, or namespace level, or the one requested by the user
+at Pod level. This asynchronous behavior will be analogous to the behavior seen with standard
+K8s applications when they talk to any Services. Applications desiring to have the same IPaddress
+for both ingress and egress traffic can create Headless Services in front of them and should
+avoid requesting explicit globalIP that match Headless Service backend Pods.
```

Comment:
> This bit is problematic since according to the diagram the "namespace IP" takes precedence over the ingress IP, but what if as a user I want the specific pods for a service to use the same address? Then I can't even have namespace level assignment that catches all other pods (or I have to use some excluding selector or something?)
> 
> As a user I think it would be simpler is the order or precedence for egress was:
> 
> 1. `GlobalEgressIP` with pod selector that matches the pod's IP
> 2. `GlobalIngressIP` (or whatever for headless service)
> 3. `GlobalEgressIP` with empty selector
> 4. `ClusterGlobalEgressIP`

Correct @mkolesnik. As described textually, this was the idea we had in mind, but I forgot to update the diagram. Thanks.
 
> **OTOH for simplicity's sake, and if we want to mimic K8s as close as possible, we should do the following:**

True, this will be simpler to support. But this was a requirement from external network use-cases @mkimuram @mangelajo 
 
> * Headless service IPs always take precedence (matches [1] AFAICT)
> * Cluster service IPs are only used for ingress
> 
> [1] https://kubernetes.io/docs/concepts/services-networking/service/#headless-services
> 
> > For headless Services, a cluster IP is not allocated, kube-proxy does not handle these Services, and there is no load balancing or proxying done by the platform for them. How DNS is automatically configured depends on whether the Service has selectors defined:

Reply:
> True, this will be simpler to support. But this was a requirement from external network use-cases @mkimuram @mangelajo

Is the requirement documented here (in the EP or a discussion)? If so I'd like to have a read to understand why it's needed.

To me this simple approach would be best since it follows K8s and also is very simple to understand and work with, so understanding the external networking requirement is important.

Reply:
@mkolesnik yes it was discussed already in this thread. Please see - https://github.com/submariner-io/enhancements/pull/22#discussion_r571030250

Reply:
> > But this was a requirement from external network use-cases
> 
> Inside a single cluster, they can use namespace selector or pod selector, howerver outside the cluster, they need to rely on ingress IPs.

We have a separate project to tackle cross cluster network policy and it'll support such use cases in an abstract way, so we shouldn't base our design on that.

Anyhow, if the effective order is going to be that `GlobalIngressIP` takes precedence over `GlobalEgressIP` with empty selector I think we can live with that, even though I don't agree that it's the best solution to the problem that has been presented.

Reply:
Thanks @mkolesnik and @mkimuram